Added a service Client interface and renamed the server interface

abjeni opened the pull request:
Example client interface:

```go
// Raft is the client-side API for the Raft Service
type RaftClient interface {
	RequestVote(ctx context.Context, in *raftpb.RequestVoteRequest) (resp *raftpb.RequestVoteResponse, err error)
	AppendEntries(ctx context.Context, in *raftpb.AppendEntriesRequest) (resp *raftpb.AppendEntriesResponse, err error)
	AppendEntries2(ctx context.Context, in *raftpb.AppendEntriesRequest) (resp *raftpb.AppendEntriesResponse, err error)
	InstallSnapshot(ctx context.Context, in *commonpb.Snapshot) (resp *raftpb.InstallSnapshotResponse, err error)
	CatchMeUp(ctx context.Context, in *raftpb.CatchMeUpRequest) (resp *raftpb.Empty, err error)
}
```

The server interface would now look like this:

```go
// Raft is the server-side API for the Raft Service
type RaftServer interface {
	RequestVote(ctx gorums.ServerCtx, request *raftpb.RequestVoteRequest) (response *raftpb.RequestVoteResponse, err error)
	AppendEntries(ctx gorums.ServerCtx, request *raftpb.AppendEntriesRequest) (response *raftpb.AppendEntriesResponse, err error)
	AppendEntries2(ctx gorums.ServerCtx, request *raftpb.AppendEntriesRequest) (response *raftpb.AppendEntriesResponse, err error)
	InstallSnapshot(ctx gorums.ServerCtx, request *commonpb.Snapshot) (response *raftpb.InstallSnapshotResponse, err error)
	CatchMeUp(ctx gorums.ServerCtx, request *raftpb.CatchMeUpRequest) (response *raftpb.Empty, err error)
}
```

Fixes #178

meling reviewed (changes requested):
Nice work!

There is a small snag though. The client interface must match the client-side implementation (in the generated `_gorums.pb.go` files); specifically, the multicast and unicast types has an extra variadic `opts` argument.

One way to enforce interface compliance is to declare a variable like this (could be in a separate file in the `dev` folder):
```go
// enforce interface compliance
var _ ZorumsServiceClient = (*Configuration)(nil)
```
This will give a compile error if the generated `Configuration` object does not match the `ZorumsServiceClient` interface.

abjeni commented:
I had to create two client interfaces, one for Configuration and one for Node, which looks like this:

```go
// Raft is the client-side Configuration API for the Raft Service
type RaftConfigurationClient interface {
	RequestVote(ctx context.Context, in *raftpb.RequestVoteRequest) (resp *raftpb.RequestVoteResponse, err error)
	AppendEntries(ctx context.Context, in *raftpb.AppendEntriesRequest, f func(*raftpb.AppendEntriesRequest, uint32) *raftpb.AppendEntriesRequest) (resp *raftpb.AppendEntriesQFResponse, err error)
}

// enforce interface compliance
var _ RaftConfigurationClient = (*Configuration)(nil)

// Raft is the client-side Node API for the Raft Service
type RaftNodeClient interface {
	AppendEntries2(ctx context.Context, in *raftpb.AppendEntriesRequest) (resp *raftpb.AppendEntriesResponse, err error)
	InstallSnapshot(ctx context.Context, in *commonpb.Snapshot) (resp *raftpb.InstallSnapshotResponse, err error)
	CatchMeUp(ctx context.Context, in *raftpb.CatchMeUpRequest) (resp *raftpb.Empty, err error)
}

// enforce interface compliance
var _ RaftNodeClient = (*Node)(nil)
```

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,60 @@
+
+var clientNodeInterface = `
+{{$genFile := .GenFile}}
+{{range .Services -}}
```
See below for func definitions; use like this:
```go
{{- range nodeOnlyServices .Services}}
```

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,60 @@
+{{$service := .GoName}}
+// {{$service}} is the client-side Node API for the {{$service}} Service
+type {{$service}}NodeClient interface {
+	{{- range .Methods}}
```
And this:
```go
{{- range nodeMethods .Methods}}
```

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,60 @@
+
+var clientConfigurationInterface = `
+{{$genFile := .GenFile}}
+{{range .Services -}}
```
```go
{{- range gorumsServices .Services}}
```

meling replied:
(see implementations of the helper funcs below)

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,60 @@
+{{$service := .GoName}}
+// {{$service}} is the client-side Configuration API for the {{$service}} Service
+type {{$service}}ConfigurationClient interface {
+	{{- range .Methods}}
```
```go
{{- range gorumsMethods .Methods}}
```

meling commented on cmd/protoc-gen-gorums/gengorums/gorums_func_map.go:
```diff
@@ -106,6 +109,12 @@ var funcMap = template.FuncMap{
+	"isConfigurationCall": func(method *protogen.Method) bool {
+		return hasMethodOption(method, gorums.E_Multicast, gorums.E_Quorumcall, gorums.E_Correctable, gorums.E_Async)
+	},
 	"methods": func(services []*protogen.Service) (methods []*protogen.Method) {
```
Perhaps we can remove this one; it does not appear to be used. I don't remember why I added it, and it seems we are using `.Methods` instead.

meling commented on benchmark/benchmark_gorums.pb.go:
```diff
@@ -170,6 +170,28 @@ func (c *Configuration) AsyncQuorumCall(ctx context.Context, in *Echo) *AsyncEch
+var _ BenchmarkConfigurationClient = (*Configuration)(nil)
+
+// Benchmark is the client-side Node API for the Benchmark Service
+type BenchmarkNodeClient interface {
```
We should avoid generating empty interfaces, and the corresponding interface compliance check if it is empty. See other comments for suggestions.

meling commented on cmd/protoc-gen-gorums/dev/zorums_client_gorums.pb.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+// ZorumsService is the client-side Configuration API for the ZorumsService Service
+type ZorumsServiceConfigurationClient interface {
```
I wonder if we want such a long name... What do you think about just calling it `ZorumsServiceClient` to align with gRPC's naming?

We can still use `ZorumsServiceNodeClient` since that's more of a special case for Gorums.

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,60 @@
+	{{- range .Methods}}
+		{{- if isConfigurationCall .}}
+			{{- if isOneway .}}
+				{{- $customOut := customOut $genFile .}}
```
This one does not appear to be needed since `$customOut` isn't used in this if-branch.

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,60 @@
+	{{- range .Methods}}
+		{{- if not (isConfigurationCall .)}}
+			{{- if isOneway .}}
+				{{- $customOut := customOut $genFile .}}
```
This one does not appear to be needed since `$customOut` isn't used in this if-branch.

meling reviewed (changes requested):
Nice work again... We should avoid creating empty interfaces for both Node and Configuration.

abjeni commented:
the qspecServices does not do what it was supposed to do. I decided to remove it since we need a QuorumSpec interface for the Configuration struct anyway.

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,105 @@
+
+var client = clientVariables + clientConfigurationInterface + clientNodeInterface
+
+// gorumsMethods returns all Gorums-specific methods, such as multicast, quorumcall, correctable, and async methods.
```
gorumsMethods --> configurationMethods

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,105 @@
+	return s
+}
+
+// nodeOnlyServices returns all node-only services, such as services with only unicast and plain gRPC methods.
```
nodeOnlyServices? Or nodeServices.

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,105 @@
+	return s
+}
+
+// gorumsServices returns all services that have Gorums methods.
```
configurationServices

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,105 @@
+{{- range configurationsServices .Services}}
+	{{$service := .GoName}}
+	// {{$service}} is the client-side Configuration API for the {{$service}} Service
+	type {{$service}}ConfigurationClient interface {
```
```go
// {{$service}}Client is the client interface for the {{$service}} service.
type {{$service}}Client interface {
```
I feel that having the `Configuration` name be part of the interface name gives it a connotation to something we can configure.

meling commented on cmd/protoc-gen-gorums/gengorums/template_client.go:
```diff
@@ -0,0 +1,105 @@
+{{$genFile := .GenFile}}
+{{- range nodeServices .Services}}
+{{$service := .GoName}}
+// {{$service}} is the client-side Node API for the {{$service}} Service
```
```go
// {{$service}}NodeClient is the single node client interface for the {{$service}} service.
type {{$service}}NodeClient interface {
```
(only comment change)

meling reviewed (changes requested):
A few more changes please. Mostly doc comment improvements and interface naming. And recall to regenerate.

abjeni commented:
I rewrote some of the comments and fixed the qspec template, now you could maybe even have multiple services without quorum calls.
renamed ServiceConfigurationClient to ConfigurationClient.
